### ISSUE TITLE
Return packet if unpack fails

### ISSUE THREAD

**Issue by matje** [miekg/dns#605]
Right now, go dns returns only the packet if it can unpack it or if the error is that the packet was truncated. However, there are other error scenarios where I would be interested in the packet bytes. 

Would you be open to always return the packet bytes in `client.ReadMsg()` regardless of what error was encountered during `msg.Unpack()`?

https://github.com/miekg/dns/blob/master/client.go#L204

```
        ...
        m := new(Msg)
        if err := m.Unpack(p); err != nil {
                return m, err
        }
        ...
```

**Comment by miekg** [comment]
[ Quoting <notifications@github.com> in "[miekg/dns] Return packet if unpack..." ]
>Right now, go dns returns only the packet if it can unpack it or if the error is that the packet was truncated. However, there are other error scenarios where I would be interested in the packet bytes.
>
>Would you be open to always return the packet bytes in `client.ReadMsg()` regardless of what error was encountered during `msg.Unpack()`?

Sure, we can do this. But your example code below doesn't return the bytes, it 
returns whatever was set in m.

I wonder what use cases you have?

/Miek

--
Miek Gieben

**Comment by matje** [comment]
Sorry for the confusion, I want to return the packet (preferably the structure, but bytes will suffice too) to check if the response is as expected. We use go dns a lot in our test suite to ensure that our server is responding correctly. For example, if our server should return a FORMERR response we would like to test that the correct Message ID, RCODE, Flags, etc are set. Right now we cannot with go dns because it will not give us back the packet, just an error (in case of a bad QNAME for example).

**Comment by matje** [comment]
A partial filled in structure `m` is better than none :)

**Comment by miekg** [comment]
(Also see #/604 for another corner case).

I'm not opposed against changing this, please send a PR. Also needs some documentation changes and no guarantees that `m` is useful, etc.

**Comment by matje** [comment]
Okay will do! Thanks

**Comment by matje** [comment]
Resolved per #606
